subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleAllprojectsIntegrationTest.groovy

```diff
@@ -0,0 +1,195 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.invocation
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+
+class GradleLifecycleAllprojectsIntegrationTest extends AbstractIntegrationSpec {
+
+    def 'gradle.lifecycle.allprojects is applying before gradle.#api'() {
+        given:
+        settingsFile << """
+            rootProject.name = 'root'
+            gradle.lifecycle.allprojects { project ->
+                println "lifecycle.allprojects for \${project.name}"
+            }
+            gradle.$invocation { project ->
+                println "$api for \${project.name}"
+            }
+            include(":a")
+        """
+
+        file("a/build.gradle") << "println ':a'"
+
+        buildFile << """
+            println 'Root'
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains expectedOutput
+        where:
+        api           | invocation    | expectedOutput
+        "allprojects" | "allprojects" | gradleAllprojectsExpectedOutput
+        "rootProject" | "rootProject" | gradleRootProjectExpectedOutput
+    }
+
+    private static def gradleAllprojectsExpectedOutput = """
+lifecycle.allprojects for root
+allprojects for root
+lifecycle.allprojects for a
+allprojects for a
+Root
+:a
+"""
+    private static def gradleRootProjectExpectedOutput = """
+lifecycle.allprojects for root
+rootProject for root
+Root
+"""
+
+    def 'gradle.lifecycle.allprojects is applying only once'() {
+        given:
+        settingsFile << """
+            rootProject.name = 'root'
+            gradle.lifecycle.allprojects { project ->
+                println "lifecycle.allprojects for \${project.name}"
+            }
+            include(":a")
+        """
+
+        file("a/build.gradle") << "println ':a'"
+
+        buildFile << """
+            println 'Root'
+            allprojects { project ->
+                println "allprojects for \${project.name}"
+            }
+            subprojects { project ->
+                println "subprojects for \${project.name}"
+            }
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains """
+lifecycle.allprojects for root
+Root
+allprojects for root
+lifecycle.allprojects for a
+allprojects for a
+subprojects for a
+:a
+"""
+    }
+
+    def 'gradle.lifecycle.allprojects is applying before gradle.lifecycle.beforeProject'() {
+        given:
+        settingsFile << """
+            rootProject.name = 'root'
+            gradle.lifecycle.allprojects { project ->
+                println "lifecycle.allprojects for \${project.name}"
+            }
+            gradle.lifecycle.beforeProject { project ->
+                println "lifecycle.beforeProject for \${project.name}"
+            }
```

Comment:
❔ Can we swap the registration order of callbacks here as well, without changing the expected test output to show that `allprojects` callbacks always take precedence?